FreeBSD12 version cleanup and process and thread enhancements

jasonpugsley opened the pull request:
This builds and runs in Release but I've seen an assert in a Debug run that I'm still investigating. I just wanted to get some early feedback.

In particular is this line too messy/lazy? I think `HAVE_PTHREAD_GETTHREADID_NP ` is currently found for MacOS but it doesn't look in `pthread_np.h` where it's defined for FreeBSD. 

src/coreclr/gc/unix/gcenv.unix.cpp line 472
`#elif HAVE_PTHREAD_GETTHREADID_NP || defined(__FreeBSD__)`


#14537

josteink commented on src/coreclr/gc/unix/gcenv.unix.cpp:
```diff
@@ -466,7 +469,7 @@ uint64_t GCToOSInterface::GetCurrentThreadIdForLogging()
 #if defined(__linux__)
     return (uint64_t)syscall(SYS_gettid);
-#elif HAVE_PTHREAD_GETTHREADID_NP
+#elif HAVE_PTHREAD_GETTHREADID_NP || defined(__FreeBSD__)
```
Wouldn't it be cleaner to add the FreeBSD check to the `HAVE_PTHREAD_GETTHREADID_NP`-check?

josteink replied:
This is currently the check being done by `src/coreclr/gc/unix/configure.cmake`:

![image](https://user-images.githubusercontent.com/411338/134003694-114f6b0e-6265-46d1-8a0e-ce49eb8d3f61.png)

I guess that fails because it lacks the relevant include. Any way to shoehorn our `#include <pthread_np.h>` into this (or a similar test)? Or just create our own similar test for the BSDs? `HAVE_PTHREAD_NP_GETTHREADID_NP`?

jasonpugsley replied:
Yeah the check needs `#include <pthread_np.h>` but that might break the OS(s) that already pass this check.

josteink replied:
I'm really not a cmake-expert, so I'm not sure what the cleanest way to solve this fairly simple problem is. :(

jasonpugsley replied:
I don't think you can define this test for `HAVE_PTHREAD_NP_GETTHREADID_NP` more than once. I'm not a CMake guru though.

josteink replied:
Note, I suggested adding a new test along-side the other one (`HAVE_PTHREAD_GETTHREADID_NP` and `HAVE_PTHREAD_NP_GETTHREADID_NP`)... And then you can define a global variable based on either of those 2 passing, something like the following quasi-code:

````c-make
SET(HAVE_PTHREAD_GETTHREADID,
    HAVE_PTHREAD_GETTHREADID_NP || HAVE_PTHREAD_NP_GETTHREADID_NP`)
```` 

If you catch my drift.

jasonpugsley replied:
Then the test goes from being
`#elif HAVE_PTHREAD_GETTHREADID_NP || defined(__FreeBSD__)`
to
`#elif HAVE_PTHREAD_GETTHREADID_NP || defined(HAVE_PTHREAD_NP_GETTHREADID_NP)`
which seems to me to be a bit confusing - and probably less clear than just `__FreeBSD__`.
I don't know - more work needed here too 😀

jasonpugsley replied:
OK, maybe not the prettiest but I think it's an improvement, I've isolated the change to the cmake file. I'm still happy to take changes for better or more canonical ways of doing this.
```cmake
if(HAVE_PTHREAD_NP_H)
  set(PTHREAD_NP_H_INCLUDE "#include <pthread_np.h>")
endif()

check_cxx_source_compiles("
    #include <pthread.h>
    ${PTHREAD_NP_H_INCLUDE}
    #include <stdint.h>
```

josteink commented on src/libraries/Native/Unix/Common/pal_error_common.h:
```diff
@@ -12,7 +12,7 @@
 
-// ENODATA is not defined in FreeBSD 10.3 but is defined in 11.0
+// ENODATA is not defined in FreeBSD
 #if defined(__FreeBSD__) & !defined(ENODATA)
```
Since we're no longer targetting these old verisons, do we even need the this guard anymore?

jasonpugsley replied:
So should I just remove `& !defined(ENODATA)` with the expectation it can never be present? I wonder if the original contributor had ENODATA defined by a package they had installed?

josteink replied:
I suspect that the original contributor was me, but I honestly don't remember why I did it this way. We can remove it if it's no longer useful (or if it never was).

josteink commented on src/libraries/Native/Unix/System.IO.Ports.Native/pal_serial.c:
```diff
@@ -13,7 +13,7 @@
+// ENODATA is not defined in FreeBSD
 #if defined(__FreeBSD__) & !defined(ENODATA)
 #define ENODATA ENOATTR
 #endif
```
Same question here. And I'm saying that as the guy who originally added this code :smile:

jasonpugsley replied:
ENODATA doesn't appear to be defined in 12.2. I haven't checked if it was actually in 11.0.

josteink replied:
If it can't be there, not point checking for it. The build is complex enough without providing new means of indirection. Let past be the past and just remove the check :smile:

jasonpugsley replied:
OK, I need to look into this a bit more then. It might be better to put the condition around the usages as is done here for `ESOCKTNOSUPPORT` if we can never produce that error.

https://github.com/dotnet/runtime/blob/4d411739ce1d9db68dfe12227663479c9d71c50d/src/libraries/Native/Unix/Common/pal_error_common.h#L313-L324

jasonpugsley replied:
I've updated this as discussed. It builds fine but I'll try to run some unit tests as well.

josteink commented:
Looks like a nice cleanup, but I'm suspecting we could go even further. See comments :)

jasonpugsley commented:
The change to remove the freebsd11 rid breaks the build `Please specify UpdateRuntimeFiles=true to commit the changes.` How do I fix that?

jasonpugsley committed:
Set FreeBSD 12 as minimum version

This is a small clean-up to remove reliance on FreeBSD versions
prior to 12 and more specifically 12.2 which is the lowest currently
supported version of FreeBSD. 11.4 becomes unsupported as of
October 2021. See https://www.freebsd.org/releases/

jasonpugsley committed:
Add functionality for FreeBSD process and threads

Added support for naming native threads, the process disambiguation
key and detecting if the process is running under a debugger.

Changed references to thr_self to pthread_getthreadid_np already
used elsewhere - it returns the same value.

jasonpugsley commented:
After re-baselining and building, running clr.tests and using a Debug build as the runtime in .dotnet to do another build everything looks good - apart from the failing checks in CI. I'm not definite that they're not related but I don't know enough to know 😂
It's ready for review.

josteink commented:
First of all, I think the changes done so far makes the end result much cleaner and nicer than it used to be. Good job!

If I were to be a "hard" critic, I would argue that the PR changes several things at once:

* it updates public documentation
* it implements code and fixes for current FreeBSD
* it obsoletes/removes support for obsolete FreeBSD versions.

It might be easier to digest and review in a meaningful way if those things were done as three different PRs, but on the flip side, the current code there does look fairly straight-forward and risk-free, particularly for existing, supported platforms.

jasonpugsley commented:
> It might be easier to digest and review in a meaningful way if those things were done as three different PRs, but on the flip side, the current code there does look fairly straight-forward and risk-free, particularly for existing, supported platforms.

I put up 2 commits - one with the FreeBSD11 obsolescence (documentation and code) and one with a set of fixes that go together (to identify processes and threads - only code).

I could break the documentation out into a separate commit but it's only 2 lines and it is solely related to removing support for FreeBSD11 - hence putting it with the code/config that changed with it.

If that doesn't make sense then I probably do need to rejig it so let me know.

josteink commented:
No biggie. Just general advice. Let's see what the official .net crew has to say about it :smile:

wfurt commented on src/libraries/Microsoft.NETCore.Platforms/src/runtimeGroups.props:
```diff
@@ -162,7 +162,7 @@
       <Parent>unix</Parent>
       <Architectures>x64</Architectures>
-      <Versions>11;12;13</Versions>
+      <Versions>12;13</Versions>
```
AFAIK we never remove old versions. It may not matter since we never shipped.

jasonpugsley replied:
Yes, we still don't have an SDK out in the wild for FreeBSD so I thought cleaning out as much old stuff will make future work easier - less technical debt too.
Let me know if I have to revert it though.

wfurt reviewed (approved):
LGTM. 
The only change I'm not sure about is the RID.

wfurt commented:
test failures looks unrelated.

jasonpugsley commented:
Thankyou!

wfurt commented:
> Thankyou!

I did not do the work :)
